### PR TITLE
394 Log errors to Slack / Logging improvement

### LINKED ISSUE #394: Implement logging to slack
See https://serafin.io/slack-django-errors/, do not use the mentioned integrations there, instead use a Slack app (ask @maltezacharias for credentials) to send the message (https://api.slack.com/tutorials/slack-apps-hello-world)

- Include logging of errors to slack devs Team 
- Include rate limiting, so that repeating errors will not clutter the channel
- Consider using a new channel (e.g. dev-logging) to be able to selectively mute it if needed
- Add Hook URL to 
- Add to startup checks #393, see if there is a sandbox like method to validate hook URL, if not send a message to the channel during system checks and verify the HTTP status code. Double check that using a wrong Hook URL will fail the system check
- Consider using https://github.com/wkeeling/ratelimitingfilter or something similar to limit amount of messages

### PR DESCRIPTION
**Pull request by maltezacharias**
# Changes

- Adds a new Custom Log Handler to send Errors to Slack
- Adds a new Environment Variable to configure Slack Webhook URL
- Overhauls the complete logging configuration
- Adds JSON structured logging format
- Adds applicationwide logging configuration (using logging.getLogger(__name__) in all project folders under apps/
- Hides django.server logs (only produced by runserver and not necessary)
- Includes new logging example in ineedstudent/views.py
- Logging to console is reduced to one line per log entry, newlines are removed, timestamps are added
- closes #394 

JSON Logging compatible with https://github.com/gistia/json-log-viewer recommended as log viewer

## TODOs: 

- [x] Password fields need to be excluded from logging (cave: request.POST!, cave: request.user!)
- [ ] Filesize should be monitored in the beginning
- [x] Care must be taken not to include data with circular references
- [ ] Consider alternate JSON lib
- [x] Performance should be monitored, increased audit logging should be added throughout the app.

## Example output:
image
image

```json
{
   "status_code":500,
   "request":{
      "method":"POST",
      "GET":{

      },
      "POST":{
         "csrfmiddlewaretoken":"",
         "name_first":"dfgs",
         "name_last":"dfgs",
         "email":"fgsfg@srdgsfghsas.sdfad",
         "phone_number":"",
         "plz":"99999",
         "countrycode":"DE",
         "umkreis":"1",
         "availability_start":"2020-04-11",
         "initial-availability_start":"2020-04-11 21:34:27",
         "braucht_bezahlung":"2",
         "zeitliche_verfuegbarkeit":"1",
         "ausbildung_typ_medstud_abschnitt":"0",
         "ausbildung_typ_mfa_abschnitt":"0",
         "ausbildung_typ_mtla_abschnitt":"0",
         "ausbildung_typ_mta_abschnitt":"0",
         "ausbildung_typ_ota_abschnitt":"0",
         "ausbildung_typ_ata_abschnitt":"0",
         "ausbildung_typ_notfallsani_abschnitt":"0",
         "ausbildung_typ_pflege_abschnitt":"0",
         "ausbildung_typ_ergotherapie_abschnitt":"0",
         "ausbildung_typ_psycho_abschnitt":"0",
         "ausbildung_typ_zahni_abschnitt":"0",
         "ausbildung_typ_physio_abschnitt":"0",
         "ausbildung_typ_kinderbetreung_ausgebildet_abschnitt":"0",
         "sonstige_qualifikationen":"keine"
      },
      "path":"/accounts/signup_student"
   },
   "level":"error",
   "message":"[django.request]: Internal Server Error: /accounts/signup_student",
   "timestamp":"2020-04-12T21:54:39.284234",
   "user":"AnonymousUser",
   "exc_info":"Traceback (most recent call last):..... ypeError: can only concatenate str (not \"__proxy__\") to str"
}
```

### CODE REVIEW

**Comment by maltezacharias** [comment]
Passwords are now hidden when DEBUG is False using the default Django Implementation to hide sensitive values from requests marked using the @method_decorator(sensitive_post_parameters())

(This is true for the default login forms)

**Review by Baschdl — CHANGES_REQUESTED** [review]
A check like in #402 would also be nice here.

Do you intend to implement the TODOs?

File: backend/match4healthcare/logging/formatters.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    ERROR = 'error'
+    CRITICAL = 'error'
+
+
+class OneLineExceptionFormatter(logging.Formatter):
+    def formatException(self, exc_info):
+        """
+        Format an exception so that it prints on a single line.
+        """
+        result = super(OneLineExceptionFormatter, self).formatException(exc_info)
```
Comment (Baschdl) [inline backend/match4healthcare/logging/formatters.py:20]: Does this output the JSON or the old format in one line?

Comment (maltezacharias) [reply backend/match4healthcare/logging/formatters.py:20]: Neither :) Basically you configure the log string in the logging configuration. The only difference is, that the default formatting will add Exception Stacktraces after your  formatted log message, this one will append it to your logline and remove newlines from the log message, putting it all on one line (**edited**)

File: backend/requirements.txt [hunk 2]
```diff
@@ -10,4 +10,6 @@ djangorestframework==3.11.0
 djangorestframework-filters==0.11.1
 whitenoise==5.0.1
 numpy==1.18.2
-pandas==1.0.3
\ No newline at end of file
+pandas==1.0.3
+requests
+JSON-log-formatter
```
Comment (Baschdl) [inline backend/requirements.txt]: Please specify the used version

**Review by Baschdl — DISMISSED** [review]

File: backend/requirements.txt (hunk 2 above)
Comment (feeds) [reply backend/requirements.txt]: done

**Review by feeds — APPROVED** [review]
This worked very nicely, maybe we should make it more transparent which messages should be logged where though.

**Pull request merged**